torch/csrc/distributed/rpc/profiler/remote_profiler_manager.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <c10/util/Optional.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/distributed/rpc/types.h>
+#include <mutex>
+#include <unordered_map>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+class TORCH_API RemoteProfilerManager {
+ public:
+  // Retrieves the lazily-initialized RemoteProfilerManager singleton instance.
+  static RemoteProfilerManager& getInstance();
+  // Sets the current, thread-local profiling key.
+  void setCurrentKey(const std::string key);
+  // Unsets the current, thread-local profiling key to allow other RPCs to reset
+  // it.
+  void unsetCurrentKey();
+  // inserts a pair (globallyUniqueId, key) to an in-memory map. The
+  // corresponding ID is used in RPC deserialization to prefix remotely profiled
+  // events with the right key.
+  void saveRPCKey(
+      GloballyUniqueId globallyUniqueId,
+      std::string rpcProfilingKey);
+  // Retrieves the profiling key corresponding to the given globallyUniqueId.
+  // Throws if it is not found.
+  std::string retrieveRPCProfilingKey(GloballyUniqueId globallyUniqueId);
+  // Generates the next unique globally unique ID.
+  local_id_t getNextLocalId();
+  // Retrieves the currently set thread-local profiling key. Throws if it is not
+  // set.
+  std::string getCurrentProfilingKey();
+  // erases the globallyUniqueId from the map. This can help save memory in the
+  // case that many RPCs are being profiled.
+  void eraseKey(GloballyUniqueId globallyUniqueId);
+
+ private:
+  RemoteProfilerManager();
+  ~RemoteProfilerManager() = default;
```

Comment:
is this intentional to delete dtor to prevent stack allocation?

Reply:
Yes, to ensure that it is only created on the heap, and we always only have a single instance of it. It is similar to the python rpc handler and dist autograd container singleton instances.